File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>

using namespace std;

bool visited[11][11][4];

int dx[] = {0, 1, 0, -1};
int dy[] = {-1, 0, 1, 0};

int to_dir(char dir) {
    if (dir == 'U') 
        return 0;
    else if (dir == 'R') 
        return 1;
    else if (dir == 'D') 
        return 2;
    else 
        return 3;
    return -1;
}

bool is_not_valid(int x, int y) {
    return x < 0 || y < 0 || x > 10 || y > 10;
}

int opposite_direction(int dir) {
    return (dir + 2) % 4;
}

int solution(string dirs) {
    int answer = 0;
    int x = 5, y = 5;
    for (auto c : dirs) {
        int dir = to_dir(c);
        int nx = x + dx[dir];
        int ny = y + dy[dir];
        
        if (is_not_valid(nx, ny)) {
            continue;
        }
        
        if (visited[y][x][dir] == false) {
            visited[y][x][dir] = true;
            visited[ny][nx][opposite_direction(dir)] = true;
            answer++;
        }
        
        x = nx;
        y = ny;
    }
    
    return answer;
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    
    private static final HashMap<Character, int[]> location = new HashMap<>();
    private static int x = 5, y = 5;
    
    static {
        location.put('U', new int[] {0, 1});
        location.put('D', new int[] {0, -1});
        location.put('L', new int[] {-1, 0});
        location.put('R', new int[] {1, 0});
    }
    
    private static boolean isValid(int nx, int ny) {
        return nx >= 0 && nx <= 10 && ny >= 0 && ny <= 10;
    }
    
    public int solution(String dirs) {
        HashSet<String> visitedPaths = new HashSet<>();
        for (char dir : dirs.toCharArray()) {
            int[] offset = location.get(dir);
            int nx = x + offset[0];
            int ny = y + offset[1];
            if (!isValid(nx, ny)) {
                continue;
            }

            visitedPaths.add(String.format("%d %d %d %d", x, y, nx, ny));
            visitedPaths.add(String.format("%d %d %d %d", nx, ny, x, y));
    
            x = nx;
            y = ny;
        }
        
        return visitedPaths.size() / 2;
    }
}
